openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,236 @@
+var _ = admission.MutationInterface(&managementCPUsOverride{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &managementCPUsOverride{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister corev1listers.NamespaceLister
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving namespace: %v", PluginName, err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isNamespaceAllowed(ns.Annotations) {
+		return nil
+	}
+
+	if !isPodManaged(pod.Labels) {
+		return nil
+	}
+
+	allContainers := append([]coreapi.Container{}, pod.Spec.InitContainers...)
+	allContainers = append(allContainers, pod.Spec.Containers...)
+	podQoSClass := getPodQoSClass(allContainers)
+
+	// we do not want to change guaranteed pods resource allocation, because it should be managed by
+	// relevant managers(CPU and memory) under the kubelet
+	if podQoSClass == coreapi.PodQOSGuaranteed {
+		klog.Warningf("Skip pod %q CPUs requests modifications because it has guaranteed QoS class", pod.Name)
+		return nil
+	}
+
+	podCopy := pod.DeepCopy()
+	updateContainerResources(podCopy, podQoSClass)
+
+	allContainers = append([]coreapi.Container{}, podCopy.Spec.InitContainers...)
+	allContainers = append(allContainers, podCopy.Spec.Containers...)
+	podQoSClassAfterModification := getPodQoSClass(allContainers)
+
+	if podQoSClass != podQoSClassAfterModification {
+		klog.Warningf("Skip pod %q CPUs requests modifications because it will change the pod QoS class from %s to %s", pod.Name, podQoSClass, podQoSClassAfterModification)
+		return nil
+	}
+
+	updateContainerResources(pod, podQoSClass)
+
+	return nil
+}
+
+func updateContainerResources(pod *coreapi.Pod, class coreapi.PodQOSClass) {
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	for i, c := range pod.Spec.InitContainers {
```

Comment:
don't iterate over `c` (a shallow copy). This asks for programming errors. Do `c := &pod.Spec.InitContainers[i]`.

Reply:
done